src/lib/utils/alignment_buffer.h

```diff
@@ -0,0 +1,219 @@
+/*
+ * Alignment buffer helper
+ * (C) 2023 Jack Lloyd
+ *     2023 René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_ALIGNMENT_BUFFER_H_
+#define BOTAN_ALIGNMENT_BUFFER_H_
+
+#include <botan/concepts.h>
+#include <botan/internal/stl_util.h>
+
+#include <array>
+#include <optional>
+#include <span>
+
+namespace Botan {
+
+enum class AlignmentBufferFinalBlock : size_t {
+   IS_NOT_SPECIAL = 0,
+   MUST_BE_DEFERRED = 1,
+};
+
+/**
+ * @brief Alignment buffer helper
+ *
+ * Many algorithms have an intrinsic block size in which they consume input
+ * data. When streaming arbitrary data chunks to such algorithms we must store
+ * some data intermittently to honor the algorithm's alignment requirements.
+ *
+ * This helper encapsulates such an alignment buffer. The API of this class is
+ * designed to minimize user errors in the algorithm implementations. Therefore,
+ * it is strongly opinionated on its use case. Don't try to use it for anything
+ * but the described circumstance.
+ *
+ * @tparam T                     the element type of the internal buffer
+ * @tparam BLOCK_SIZE            the buffer size to use for the alignment buffer
+ * @tparam FINAL_BLOCK_STRATEGY  defines whether the final input data block is
+ *                               retained in handle_unaligned_data() and must be
+ *                               manually consumed
+ */
+template <typename T,
+          size_t BLOCK_SIZE,
+          AlignmentBufferFinalBlock FINAL_BLOCK_STRATEGY = AlignmentBufferFinalBlock::IS_NOT_SPECIAL>
+   requires(BLOCK_SIZE > 0)
+class AlignmentBuffer {
+   public:
+      AlignmentBuffer() : m_position(0) {}
+
+      ~AlignmentBuffer() { clear(); }
+
+      AlignmentBuffer(const AlignmentBuffer& other) = default;
+      AlignmentBuffer(AlignmentBuffer&& other) noexcept = default;
+      AlignmentBuffer& operator=(const AlignmentBuffer& other) = default;
+      AlignmentBuffer& operator=(AlignmentBuffer&& other) noexcept = default;
+
+      void clear() {
+         clear_mem(m_buffer.data(), m_buffer.size());
```

Comment:
Should be `secure_scrub_memory` to prevent being elided during the destructor.

Reply:
(Or alternately continue to use `clear_mem` here and just invoke `secure_scrub_memory` once, in the destructor)